packages/vue-app/template/client.js

```diff
@@ -102,31 +102,29 @@ function componentOption (component, key, ...args) {
   return option
 }
 
-function mapTransitions (Components, to, from) {
+function mapTransitions (toComponents, to, from) {
   const componentTransitions = (component) => {
     const transition = componentOption(component, 'transition', to, from) || {}
     return (typeof transition === 'string' ? { name: transition } : transition)
   }
-  const tComponents = [].concat(Components)
-
-  // If leaving a child route to a parent, keep the child leave transition
-  while (from && from.matched.length > tComponents.length) { // eslint-disable-line no-unmodified-loop-condition
-    tComponents.push({})
+  
+  const fromComponents = from ? getMatchedComponents(from) : []
+  const maxDepth = Math.max(toComponents.length, fromComponents.length)
+
+  const mergedTransitions = []
+  for (let i=0; i<maxDepth; i++) {
+    // Clone original objects to prevent overrides
+    const toTransitions = Object.assign({}, componentTransitions(toComponents[i]))
+    const transitions = Object.assign({}, componentTransitions(fromComponents[i]))
+    
+    // Combine transitions & prefer `enter` callbacks and name of "to" route
+    Object.keys(toTransitions)
+        .filter(key => typeof toTransitions[key] !== 'undefined' && !key.toLowerCase().includes('leave'))
```

Comment:
Reversed the logic so that we start from the "from" properties and merge only existing "to" properties, ignoring the leave properties as before. This allows to keep the transition name from the "from" component in case there is no "to" component at the same depth ( as it happens when going from child to parent ).

Also just checking for truthiness of `toTransitions[key]` was preventing falsy values from being merged, ie: `css: false`, so `typeof toTransitions[key] !== 'undefined'` fixes that as well.